### ISSUE TITLE
R 4.4 + 4.3 Win UCRT migration: recurring issues

### ISSUE THREAD

**Issue by mbargull** [conda-forge/r-base-feedstock#325]
### Comment:

We are starting to see some recurring issues in the migration.
Opening a list here for visibility:

-   [ ] `win-64`, `r-base=4.4`: `.../x86_64-w64-mingw32/bin/ld.exe: cannot find -lgfortran: No such file or directory`
    Only affects the `4.4` builds but not the `4.3` although the environments are completely identical apart from R packages!
    Seen, e.g., in
        - https://github.com/conda-forge/r-matrix-feedstock/pull/44
        - https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=966011&view=logs&j=71a23920-152a-53ca-00d8-b45290d07ed2&t=a0e11902-fa51-5ad4-87d2-f8f2ba048f22&l=2455

-   [ ] `win-64`, `r-base=4.3|4.4`: Abrupt failure after `*** moving datasets to lazyload DB` with no further log output.
    Seen, e.g., in
        -   https://github.com/conda-forge/r-urltools-feedstock/pull/17
        -   https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=966414&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=623c2ba8-c4f1-582d-6c78-1c7699e2e0a2&l=915

    . We had already seen this in
        -   https://github.com/conda-forge/r-base-feedstock/pull/278#issuecomment-2174585455 too,
        -   https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=956445&view=logs&j=8af117b0-dfa5-5d1c-b02e-a448dd16efdb&t=975b9034-3d9e-51af-5c2e-956ac2fc1d58&l=4155

    but it did not resurface once `nlme` has been built separately in its feedstock.

**Comment by isuruf** [comment]
First issue is probably fixed by https://github.com/conda-forge/r-base-feedstock/pull/326. Needs a backport to 4.3 too after confirming that it is fixed.

**Comment by isuruf** [comment]
Third issue is fixed via https://github.com/conda-forge/pkg-config-feedstock/pull/42

**Comment by mfansler** [comment]
Actually, setting PKG_CONFIG_PATH didn't "just work" for the `r-terra` build - still troubleshooting why it's not being found.

**Comment by mbargull** [comment]
I debugged
> `win-64`, `r-base=4.3|4.4`: Abrupt failure after `*** moving datasets to lazyload DB` with no further log output.

and apparently this is due to `bzfile` not working -- at least for the `urltools` case I looked at.
`gzfile` can read `.gz`/`.xz` no problem, but fails for `.bz2`.
Some packages seem to have `.rda` files that are bzip2 compressed, e.g., https://github.com/Ironholds/urltools/blob/ccdd1b4adefe6743fb88f4f2b078811d75de0c9c/data/suffix_dataset.rda .

**Comment by mfansler** [comment]
> Third issue is fixed via [conda-forge/pkg-config-feedstock#42](https://github.com/conda-forge/pkg-config-feedstock/pull/42)

This still needs an `m2-pkg-config` build to work for us, correct? E.g., see https://github.com/conda-forge/r-httpuv-feedstock/pull/43

**Comment by mfansler** [comment]
I've had mixed results getting UCRT + `pkg-config` working. The boilerplate Makevars.ucrt that many R packages use wants `pkg-config` to be able to resolve all dependencies of dependencies and unfortunately not all our Windows packages ship with a .pc. Otherwise, the script falls back to giving flags for a long list of libraries, most which are unnecessary (I'm guessing it's every library they default-install in the CRAN Win runners).

Instead, I find it simpler to gut the Makevars.ucrt with a patch to list only libraries needed to link. For example, with the `r-tiff` package, I patch Mavevars.ucrt to simply:

```bash
PKG_LIBS = -L$(LIBRARY_LIB) -ltiff
PKG_CPPFLAGS = -I$(LIBRARY_INC)
```

It's manual labor, but straightforward.

**Comment by mfansler** [comment]
Any thoughts about adding `-pthread` somewhere to the `r-base` Makeconf for our Windows builds?

**Comment by danielnachun** [comment]
I've manually reviewed all of the "In PR" feedstocks and identified some low hanging fruit we can fix relatively easily.

### Missing DSO whitelist
These packages are missing some `.dll` files from the `missing_dso_whitelist`:
- https://github.com/conda-forge/r-spiderbar-feedstock/pull/1 - `R.dll`
- https://github.com/conda-forge/r-minty-feedstock/pull/1 - `R.dll`
- https://github.com/conda-forge/r-box-feedstock/pull/3 - `R.dll`
- https://github.com/conda-forge/r-gmwt-feedstock/pull/2 - `R.dll`, `Rblas.dll`
### Missing host requirement
These packages are missing at least one library 
- https://github.com/conda-forge/r-idefix-feedstock/pull/1 - `libgomp`
- https://github.com/conda-forge/r-bart-feedstock/pull/4 - `libgomp`
- https://github.com/conda-forge/r-rvcg-feedstock/pull/4 - `libgomp`
- https://github.com/conda-forge/r-allelicseries-feedstock/pull/4 - `libgomp`
- https://github.com/conda-forge/r-fedmatch-feedstock/pull/1 - `libgomp`
- https://github.com/conda-forge/r-glpkapi-feedstock/pull/5 - `gmp`
- https://github.com/conda-forge/r-httpgd-feedstock/pull/6 - `harfbuzz`, `freetype`, `libintl`, `cfitsio` - these may not actually be needed as some packages add a bunch of flags to `PKG_LIBS` that are only needed for static linking or other cases that don't apply to us
### Missing both DSO whitelist and host requirements
These packages are missing both some .dll files in `missing_dso_whitelist` and at least one library
- https://github.com/conda-forge/r-fdasrvf-feedstock/pull/4
  - `missing_dso_whitelist`: `R.dll`, `Rblas.dll`
  -  missing host requirement: `libgomp`
- https://github.com/conda-forge/r-jmbayes2-feedstock/pull/2
  - `missing_dso_whitelist`: `Rblas.dll`, `Rlapack.dll`
  - missing host requirement: `libgomp`
- https://github.com/conda-forge/r-stpphawkes-feedstock/pull/1
  - `missing_dso_whitelist`: `R.dll`
  -  missing host requirement: `libgomp`
- https://github.com/conda-forge/r-xslt-feedstock/pull/5
  - `missing_dso_whitelist`: `R.dll`
  - missing host requirement: `libiconv`, `zlib`
### Missing flags in `PKG_LIBS`
These packages need at least one flag to be added to `PKG_LIBS`.  In some cases they may be missing because the associated library is missing a `.pc` file for `pkg-config` as noted above.  We've also discussed specifically for `-pthread` just adding this to `Makeconf` for `r-base`:
- https://github.com/conda-forge/r-proj-feedstock/pull/8 - `-lproj`
- https://github.com/conda-forge/r-websocket-feedstock/pull/3. - `-pthread`
- https://github.com/conda-forge/r-sysfonts-feedstock/pull/16 - `-lfreetype2`
### Missing build dependency
These packages are missing at least one build dependency.  
- https://github.com/conda-forge/r-metabma-feedstock/pull/11 - `m2-grep`
- https://github.com/conda-forge/r-cometexacttest-feedstock/pull/9 - `make`
- https://github.com/conda-forge/r-destinie-feedstock/pull/8 - `make`
- https://github.com/conda-forge/r-gmd-feedstock/pull/8 - `make`

**Comment by danielnachun** [comment]
Separately from these known issues, I've seen two more recurrent issues not previous mentioned that don't currently have a solution:
### Unable to find `binutils` executables
`binutils` should already be installed by the `build` requirements, but these packages cannot find `ar` to build a static library:
- https://github.com/conda-forge/r-leidenalg-feedstock/pull/12
- https://github.com/conda-forge/r-n2r-feedstock/pull/4
- https://github.com/conda-forge/r-scs-feedstock/pull/5

In that last PR I think we determined that `ar` is installed, but the build system for the packages can't find it.
### Cannot find a path
These packages compile successfully but fail afterward with rather vague error `The system cannot find the path specified.`
- https://github.com/conda-forge/r-openmx-feedstock/pull/48
- https://github.com/conda-forge/r-ndjson-feedstock/pull/8
- https://github.com/conda-forge/r-prophet-feedstock/pull/13

**Comment by zklaus** [comment]
The issue of not finding `ar` was briefly discussed today in the conda-forge core meeting. The notes can be found [here](https://hackmd.io/@conda-forge/S1XT61jCR).

Basically,  builds need to use the `$AR` environment variable and not rely on a hard coded `ar` binary because the new packages only have prefixed versions of the program, e.g. `x86_64-w64-mingw32-ar`.

**Comment by danielnachun** [comment]
Thanks @zklaus that makes a lot of sense.  For `r-leidenalg` I was able to pretty quickly find where `ar` was hard-coded, so this should be easy to patch.  For the other two it's a little less obvious but as long as we know what to try now I think this can be resolved relatively quickly.

**Comment by danielnachun** [comment]
I've reviewed the bot errors now and will summarize some findings below.  Most of the failures are currently due to `r-sf`, `r-terra`, and `r-terra` missing on Windows, and I won't post those because we already why those failures are happening.

All of these packages are failing because the dependency `m2w64-libwinpthread-git` can't be resolved:
- https://github.com/conda-forge/r-nanonext-feedstock
- https://github.com/conda-forge/r-rxode2ll-feedstock
- https://github.com/conda-forge/r-parglm-feedstock
- https://github.com/conda-forge/r-admm-feedstock
- https://github.com/conda-forge/r-maotai-feedstock
- https://github.com/conda-forge/r-rlemon-feedstock
- https://github.com/conda-forge/r-symphony-feedstock
- https://github.com/conda-forge/r-didimputation-feedstock/
I'm wondering if `m2w64-libwinpthread-git` should be automatically removed by the migrator bot before trying to resolve dependencies.  I'm not sure we need it any more with UCRT

There is also one package that fails due to the `m2w64-toolchain` dependency can't be resolved: https://github.com/conda-forge/r-bmisc-feedstock.  I'm also wondering if that `m2w64-toolchain` should be removed by the bot.

There is also one package that fails because `m2w64-tk` and `m2w64-tcl` can't be be resolved: https://github.com/conda-forge/r-tkrplot-feedstock.  I think we now have `tk` available on Windows so we can probably use that instead.

These packages are failing because `jags` is missing on Windows:
- https://github.com/conda-forge/r-best-feedstock
- https://github.com/conda-forge/r-miscf-feedstock
- https://github.com/conda-forge/r-rjags-feedstock

These packages are failing because `gtk2` is missing on Windows:
- https://github.com/conda-forge/r-rgtk2-feedstock
- https://github.com/conda-forge/r-cairodevice-feedstock

**Comment by mfansler** [comment]
FYI, workaround for the `r-locfit` [issue of undefined C standards](https://github.com/conda-forge/r-base-feedstock/issues/349) can go something like the following in `build.sh`:

```bash
## manual fix for https://github.com/conda-forge/r-base-feedstock/issues/349
if [[ "$target_platform" == "linux-aarch64" || "$target_platform" == "linux-ppc64le" || "$target_platform" == "osx-arm64" ]]; then
    if [[ "$build_platform" != "$target_platform" ]]; then
        MAKECONF="${BUILD_PREFIX}/lib/R/etc/Makeconf"
    else
        MAKECONF="${PREFIX}/lib/R/etc/Makeconf"
    fi

    # capture CC and CFLAGS entries
    CC_VALUE=$(sed -n 's/^CC = \(.*\)/\1/p' "${MAKECONF}")
    CFLAGS_VALUE=$(sed -n 's/^CFLAGS = \(.*\)/\1/p' "${MAKECONF}")
    
    # insert into CC17 and C17FLAGS definitions
    sed -i "s|^CC17 =.*|CC17 = $CC_VALUE -std=gnu17|" "${MAKECONF}"
    sed -i "s|^C17FLAGS =.*|C17FLAGS = $CFLAGS_VALUE|" "${MAKECONF}"

    # debug
    cat ${MAKECONF}
fi
```

<sub>**NB:** This only does the C17 - so adjust accordingly.</sub>

Not sure if necessary, but for good measure I also added to `meta.yaml`:

```yaml
build:
  ## ignore workaround patching for https://github.com/conda-forge/r-base-feedstock/issues/349
  ignore_prefix_files:
    - lib/R/etc/Makeconf
```

**Comment by danielnachun** [comment]
Thanks to a tip from @mfansler, I found that essentially all of the remaining "unsolvable" packages on Windows are due to `{{ native }}` dependencies that do indeed need to be removed.  I have set up manual migrations for all of the feedstocks with this issue:

- https://github.com/conda-forge/r-rxode2ll-feedstock/pull/3
- https://github.com/conda-forge/r-parglm-feedstock/pull/1
- https://github.com/conda-forge/r-symphony-feedstock/pull/1
- https://github.com/conda-forge/r-admm-feedstock/pull/3
- https://github.com/conda-forge/r-maotai-feedstock/pull/4
- https://github.com/conda-forge/r-didimputation-feedstock/pull/5
- https://github.com/conda-forge/r-bmisc-feedstock/pull/1
- https://github.com/conda-forge/r-rlemon-feedstock/pull/2
- https://github.com/conda-forge/r-wicket-feedstock/pull/6
- https://github.com/conda-forge/r-autometric-feedstock/pull/1
- https://github.com/conda-forge/r-specsverification-feedstock/pull/5
- https://github.com/conda-forge/r-tkrplot-feedstock/pull/11

These two PRs now pass on Windows but need https://github.com/conda-forge/r-pki-feedstock/pull/23 to be merged for the dependency tree to resolve correctly for R 4.4 on linux-aarch64 and linux-ppc64le
- https://github.com/conda-forge/r-brms-feedstock/pull/40
- https://github.com/conda-forge/r-rstanarm-feedstock/pull/23

This PR needs https://github.com/conda-forge/r-lda-feedstock/pull/11 to be merged to so `r-lda` is available on Windows:
- https://github.com/conda-forge/r-stm-feedstock/pull/13

These two PRs are for packages that depend on `r-rgdal`, which is now archived because it has been removed from CRAN.  In both cases this dependency can just be removed because the packages no longer actually depend on `r-rgdal`:
- https://github.com/conda-forge/r-pgirmess-feedstock/pull/14
- https://github.com/conda-forge/r-tigris-feedstock/pull/15

The only package I've encountered now that will not be easily fixable is https://github.com/conda-forge/r-svars-feedstock, which needs `r-copula`, which we unfortunately had to merge with Windows disabled.  Unfortunately the error we saw was extremely cryptic - I will maybe try to follow up with upstream about this when I have a chance as it was not clear why the error only occurred on Windows. In case anybody wants to look into this, the error can be seen here: https://github.com/conda-forge/r-copula-feedstock/pull/21#issuecomment-2327593620.

**Comment by danielnachun** [comment]
I've also now manually migrated all the R feedstocks with bot errors, which were all due to missing license files which I added:
- https://github.com/conda-forge/r-rcsdp-feedstock/pull/17
- https://github.com/conda-forge/r-venneuler-feedstock/pull/6
- https://github.com/conda-forge/r-gapminder-feedstock/pull/9
- https://github.com/conda-forge/r-babynames-feedstock/pull/3
- https://github.com/conda-forge/r-fftrees-feedstock/pull/14
- https://github.com/conda-forge/r-rnaturalearthdata-feedstock/pull/5
- https://github.com/conda-forge/r-cetcolor-feedstock/pull/4
- https://github.com/conda-forge/r-treebase-feedstock/pull/5
- https://github.com/conda-forge/r-airr-feedstock/pull/10
- https://github.com/conda-forge/r-srs-feedstock/pull/3
- https://github.com/conda-forge/r-n1qn1-feedstock/pull/12
- https://github.com/conda-forge/r-kappalab-feedstock/pull/10

**Comment by danielnachun** [comment]
> ### Cannot find a path
> 
> These packages compile successfully but fail afterward with rather vague error `The system cannot find the path specified.`
> 
> * [[bot-automerge] Rebuild for r-base 4.4 and UCRT mingw toolchain r-openmx-feedstock#48](https://github.com/conda-forge/r-openmx-feedstock/pull/48)
> * [[bot-automerge] Rebuild for r-base 4.4 and UCRT mingw toolchain r-ndjson-feedstock#8](https://github.com/conda-forge/r-ndjson-feedstock/pull/8)
> * [[bot-automerge] Rebuild for r-base 4.4 and UCRT mingw toolchain r-prophet-feedstock#13](https://github.com/conda-forge/r-prophet-feedstock/pull/13)

Now that I have a working local Windows system I quickly figured out that all the failures are caused by these feedstocks using a now irrelevant workaround in `bld.bat`:
```
rem The system Makeconf file takes priority over Makevars.win in the src dir.
rem In conda's system Makeconf file, CXX14 is set to nothing, causing the
rem compilation to break, even when it is set properly in Makevars.win
rem So we append the proper Makevars to the system Makeconf here
set MAKECONF_FILE="%BUILD_PREFIX%\lib\R\etc\x64\Makeconf"
copy "%MAKECONF_FILE%" "%MAKECONF_FILE%.old"
echo CXX14 = $^(BINPREF^)g++ $^(M_ARCH^) >> "%MAKECONF_FILE%"
echo CXX14STD = -std=gnu++14 >> "%MAKECONF_FILE%"
echo CXX14FLAGS = -Wa,-mbig-obj -g0 -O2 >> "%MAKECONF_FILE%"

"%R%" CMD INSTALL --build .
IF %ERRORLEVEL% NEQ 0 exit 1

rem Set Makeconf back to what it was before
move /Y "%MAKECONF_FILE%.old" "%MAKECONF_FILE%"
```

While our CI is currently down in local testing I found that simply removing the workaround leads to a successful build.  So we can now rule out any sort of systematic issue with our build infrastructure as the cause of this recurring issue.

**Comment by danielnachun** [comment]
I've completed a very thorough first pass through all the low hanging fruit in the migration, and there are now only a handful of significant blockers which I will describe here below

#### R-universe packages
We currently have a 5 packages from R-universe in the migration queue.  3 of them depend on the other two:
- https://github.com/conda-forge/r-set6-feedstock
- https://github.com/conda-forge/r-param6-feedstock (needs `r-set6`)
- https://github.com/conda-forge/r-distr6-feedstock (needs `r-param6`)
- https://github.com/conda-forge/r-survivalmodels-feedstock
- https://github.com/conda-forge/r-mlr3proba-feedstock (needs `r-distr6`, `r-survivalmodels`)

I'm not really sure about the future of R-universe packages in conda-forge.  As noted in `meta.yaml` for these feedstocks, the checksum for the source archives is not stable over time and this makes automated maintenance of these packages very tedious, not to mention making it impossible to know if the archive has been compromised.  We can certainly get these migrated for now by just fixing the checksums, but we may want to discuss if they should be archived instead if they are going to be too much of a pain to maintain in the future.

#### r-terra and r-raster
I was able to make very significant progress in get `r-terra` to build on Windows in [this](https://github.com/conda-forge/r-terra-feedstock/pull/102) PR, but I've now hit a wall where the compiled package fails to load by crashing.  I'll post the exact error I get with `lldb` in the PR but it's not terribly helpful.  I think there are about 26 native dependents of either `r-terra` or `r-raster` that would need manual migration.  This is tedious but probably doable, though I don't like the idea of leaving Windows users with out a functioning version of `r-terra` and `r-raster` and unable to install any of their dependents, including the many other `noarch` packages.  Of course the better outcome here is just to fix `r-terra` but I don't know how difficult that will be.  I believe all but 4 of the "Not solvable" packages are blocked by `r-terra` or `r-raster`, as are many of the "Awaiting parents" packages as well.

#### xgboost
I think we have `xgboost` mostly working now in [this](https://github.com/conda-forge/xgboost-feedstock/pull/196) PR.  The only snag is that the Windows CUDA R package is failing its test with no obvious error.  I'm inclined to skip this test as the CPU version is identical and works fine.  `xgboost` has 15 dependents in total.

#### r-symengine
This seems to be failing because on Windows it is depending on the deprecated, archived `mpir` instead of `gmp`.  There is as [stalled PR](https://github.com/conda-forge/symengine-feedstock/pull/58) to move it to `gmp` - if we get this to work then we can migration `r-symengine` to `gmp` as well and hopefully that will get this working.  The current blocker is some hard to understand test failures on Windows - I don't know if we should just disable or skip these tests. `r-symengine` has 9 dependents in total.

#### jags
We have two "Not solvable" packages that needs `jags`, which has never had a Windows build.  I took at pass at building `jags` on Windows locally with the `autotools_clang_conda` wrapper and made some progress - I can push that so others can take a look.  If we don't want to wait for that we can manually migrate the 3 native dependents or `jags`:
- https://github.com/conda-forge/r-runjags-feedstock
- https://github.com/conda-forge/r-robma-feedstock (needs `r-runjags`)
- https://github.com/conda-forge/r-miscf-feedstock

I struck out the `gtk2` packages in my previous post because it turns out all of those packages were removed from CRAN.  The feedstocks have been archived and are now removed from the migration.

The remaining packages that are "In PR" or "Not solvable" have few/no dependents and more obscure failures.  I'll write up a summary of those later but they're not as high of a priority because they have so few dependents.

**Comment by danielnachun** [comment]
I have an important update on this issue - we have migrated every CRAN package to R 4.4 except this feedstock (and the failure has nothing to do with the migration from what I can tell): https://github.com/conda-forge/r-volumodel-feedstock/pull/2.  There are a couple of GitHub R packages that are failing or blocked, but their issues are also unrelated to the migration.

If there are no objections in the next few days, I'm going to close this issue and open a new one to track the small number of remaining holdouts.  At this point we certainly have any recurring issues given how few packages are left to migrate!

**Comment by danielnachun** [comment]
The migration is now closed.  I will open a new issue here to track the last stragglers.  Huge thanks to everyone that made this massive effort possible!
